zanata-war/src/main/java/org/zanata/dao/TextFlowTargetHistoryDAO.java

```diff
@@ -159,4 +174,136 @@ public boolean findConflictInHistory(HTextFlowTarget target,
+                        "  ) as all_translation" +
+                        "  group by " + dateOfLastChanged + ", iteration, locale, state " +
+                        "  order by lastChanged, iteration, locale, state";
+        Query query = getSession().createSQLQuery(queryString)
+                .setParameter("user", user.getId())
+                .setTimestamp("fromDate", fromDate.toDate())
+                .setTimestamp("toDate", toDate.toDate());
+        @SuppressWarnings("unchecked")
+        List<Object[]> result = query.list();
+        ImmutableList.Builder<UserTranslationMatrix> builder =
+                ImmutableList.builder();
+        for (Object[] objects : result) {
+            Date savedDate = new DateTime(objects[0]).toDate();
+            HProjectIteration iteration =
+                    loadById(objects[1], HProjectIteration.class);
+            HLocale locale = loadById(objects[2], HLocale.class);
+            ContentState savedState = ContentState.values()[(int) objects[3]];
+            long wordCount =
+                    ((BigDecimal) objects[4]).toBigInteger().longValue();
+            UserTranslationMatrix matrix =
+                    new UserTranslationMatrix(savedDate, iteration, locale,
+                            savedState, wordCount);
+            builder.add(matrix);
+        }
+        return builder.build();
+    }
+
+    @VisibleForTesting
+    @DatabaseSpecific("uses mysql function")
+    protected String convertTimeZoneFunction(String columnName,
+            Optional<DateTimeZone> userZoneOpt, DateTimeZone systemZone) {
+        if (userZoneOpt.isPresent()) {
+            String userOffset = getOffsetAsString(userZoneOpt.get());
+            String systemOffset = getOffsetAsString(systemZone);
+            return String.format("CONVERT_TZ(%s, '%s', '%s')", columnName, systemOffset, userOffset);
+        }
+        // no need to convert timezone
+        return columnName;
+    }
+
+    // This is so we can override it in test and be able to test it against h2
+    @VisibleForTesting
+    @DatabaseSpecific("uses mysql function")
+    protected String stripTimeFromDateTimeFunction(String columnName) {
+        return "date(" + columnName + ")";
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T loadById(Object object, Class<T> entityClass) {
+        return (T) getSession().byId(entityClass).load(
+                ((BigInteger) object).longValue());
+    }
+
+    private static String getOffsetAsString(DateTimeZone zone) {
+        int standardOffset = zone.getStandardOffset(0);
+        String prefix = "";
+        if (standardOffset < 0) {
+            prefix = "-";
+            standardOffset = -standardOffset;
+        }
+        return String.format("%s%02d:00", prefix,
+                TimeUnit.MILLISECONDS.toHours(standardOffset));
+    }
+
+    @Getter
+    @RequiredArgsConstructor
+    public static class UserTranslationMatrix {
```

Comment:
If this class is going to be used as a DTO, I'd rather not have it be an inner class.